### DISCUSSION TITLE
[Feature]: Standalone cli entry

### DISCUSSION THREAD

**Discussion by ahmedkhalf** [LunarVim/LunarVim#1787] [Ideas]
### Feature motivation

Right now, lvim is just an alias for `nvim -u ~/.local/share/lunarvim/lvim/init.lua --cmd "set runtimepath+=~/.local/share/lunarvim/lvim" "$@"`. But I believe it can be much more that that. 

### Describe the expected solution

Allowing lunarvim to be wrapped in an actual cli will enable endless possibilities, and will aid with the transition of lunarvim going from "just a neovim config", to a standalone IDE.

Consider the following cli entries:
```
lvim update
  update lunarvim without having to cd and git pull. (Perhaps packer sync too?)

lvim install {branch: ["master"], "rolling" }
  install lunarvim, or switch branch

lvim clean
  clean packer, cache, log.. etc. you name it.

lvim doctor
  diagnose common issues and checks for errors

lvim settings
  dump all "lvim." settings to terminal

...

lvim
  launch lunarvim, and enable restart functionality**
```

***restart functionality:*
Let's say the user sets `number` to be false, the problem is that this is a window option, so with the current hot reload implementation this will only change that specific `config.lua` window and not a buffer in a split. Furthermore, our current implementation of hot reload bypasses plugin keymappings. The solution to this is to restart neovim, currently, a user can manually restart neovim by closing and opening neovim, but then the splits aren't recovered so he will have to create a session too if he wants his splits recovered. This can be remedied/automated though, using the restart functionality.

### Describe the alternatives you've considered

There are many languages in which this app can be developed, but it seems golang is the clear winner. I tried using lua with `nvim --headless -u NONE`, but this would prove to be too slow with a whopping 20ms startup.

### Additional context

This is a small proof of concept of the restart functionality:

https://user-images.githubusercontent.com/36672196/130928696-d4360bf7-8610-4de2-9777-0a59881f5781.mp4

**Comment by Uzaaft** [comment]
It says right there in the git repo the following: 
```
An IDE layer for Neovim with sane defaults. Completely free and community driven.
```
So I don't find myself agreeing with the statement: ` will aid with the transition of lunarvim going from "just a neovim config", to a standalone IDE.`. Perhaps the other guys can pitch in regarding this.   
Also, just a definition nitpicking, what you are describing is a cli(command line interface). I was confused reading the whole "issue" before understanding that what you are talking about is a cli and not an app.

**Comment by ahmedkhalf** [comment]
I am kind of tired of people replying just for the sake of replying. This has happened not only with this issue, but other PRs in lunarvim too.
In fact it says right there in the git repo:
```
An IDE layer
```
What I want:
```
A standalone IDE
```
And my intentions were very clear too:
> to a **standalone** IDE.
> Consider the following **cli** entries:

While the title is a bit misleading (which I will rectify), the content is clear.
Please avoid useless and unnecessary comments regarding semantics, as they only hinder development. What I want is feedback regarding the idea as a whole, not regarding my use of vocabulary. Sry if I'm a bit harsh but this need to be said for future comments too.

**Comment by Uzaaft** [comment]
I believe that me pointing out the definition part was important to avoid anyone else thinking what I though, which was:
```
Huh, is Lunarvim going in the direction of Neovide?
```
I'm sorry if you experienced that as rude.

**Comment by ahmedkhalf** [comment]
No worries, I didn't think people would view it this way. This is why I have rectified the post further now. Thx for letting me know, and I apologize on my end as well.

**Comment by chase** [comment]
I think having additional functionality to update, clean, install, etc. would be very handy!

I however, do not agree about using Go. Although it does allow cross-platform building, I think it would introduce unnecessary complexity to distribute LunarVim.

[20ms is barely over the threshold of human perception of 18ms](https://news.mit.edu/2014/in-the-blink-of-an-eye-0116), so I don't think for one-off commands it is too terrible. Especially considering the majority of these operations will be doing relatively much slower, asynchronous IO.

That said, install and update are already handled by [install.sh](https://github.com/LunarVim/LunarVim/blob/rolling/utils/installer/install.sh#L287) using some simple bash scripting. The overhead of using bash is far lower for this use case.

To enhance `lvim` we could add commands using something like this proof of concept:
``` bash
#!/bin/sh

LUNARVIM_RUNTIME_DIR=${LUNARVIM_RUNTIME_DIR:-"$HOME/.local/share/lunarvim"}

for arg do
  shift
  case $arg in
    (update):
      echo 'updating'
      exit
      ;;
    (install):
      echo 'installing'
      exit
      ;;
    (*) set -- "$@" "$arg" ;;
  esac
done

exec nvim -u "$LUNARVIM_RUNTIME_DIR"/lvim/init.lua "$@"
```

This way, `lvim update` and `lvim install` will run `echo 'updating'` and `echo 'installing'` respectively, but otherwise default to running nvim as expected. In this same way we could call `$LUNARVIM_RUNTIME_DIR/lvim/utils/installer/install.sh` and other scripts as necessary.

---

Also worth noting, the majority of the overhead of starting Neovim is from setting locale, which causes a lot of context switching:
`/usr/bin/time --verbose nvim --clean --headless --cmd 'echo "hello"|q'`
```
	User time (seconds): 0.00
	System time (seconds): 0.00
	Percent of CPU this job got: 14%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:00.04
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 6568
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 51
	Minor (reclaiming a frame) page faults: 1022
	Voluntary context switches: 97
	Involuntary context switches: 0
	Swaps: 0
	File system inputs: 0
	File system outputs: 0
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```

Near realtime, high priority (no context switching):
`sudo chrt -f 99  /usr/bin/time --verbose nvim --clean --headless --cmd 'echo "hello"|q'`
```
	User time (seconds): 0.00
	System time (seconds): 0.00
	Percent of CPU this job got: 100%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:00.00
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 6380
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 293
	Voluntary context switches: 0
	Involuntary context switches: 0
	Swaps: 0
	File system inputs: 0
	File system outputs: 0
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```

**Comment by ahmedkhalf** [comment]
> I however, do not agree about using Go. Although it does allow cross-platform building, I think it would introduce unnecessary complexity to distribute LunarVim.

I fully agree with you. Right now I am actually considering building a Lua launcher using https://github.com/ers35/luastatic , using lua will make it more maintainable and building it to an executable will remove the need of having the neovim headless latency

Edit:

Also regarding modifying bash, while it works, I would actually prefer if we didn't rely on bash at all in order to support windows in the future, hence one of the reasons I am creating this launcher

**Comment by chase** [comment]
> Also regarding modifying bash, while it works, I would actually prefer if we didn't rely on bash at all in order to support windows in the future, hence one of the reasons I am creating this launcher

I think it might be worth ignoring native Windows but ensuring WSL supports LunarVim. Much less of a maintenance burden for LunarVim's maintainers, most of it is shifted onto Microsoft WSL team.

**Comment by ahmedkhalf** [comment]
> I think it might be worth ignoring native Windows but ensuring WSL supports LunarVim. Much less of a maintenance burden for LunarVim's maintainers, most of it is shifted onto Microsoft WSL team.

Sure will take this into consideration

**Comment by pidgeon777** [comment]
My vote goes to a Windows binary, just because I'm not allowed to run a WSL subsystem in my company.

Maybe other users could be in the same situation (forced to stick with a native Windows version), thus I think a poll could be a great idea, to choose between:

- Windows native binary
- WSL support
